benchmark/cluster.py

```
@@ -0,0 +1,154 @@
+    """
+    schema = {
+        "$schema": "http://json-schema.org/draft-04/schema#",
+        "type": "object",
+        "required": ["control_node", "agent_nodes"],
+        "properties": {
+            "control_node": {
+                "type": "string",
+            },
+            "agent_nodes": {
+                "type": "array",
+                "items": {
+                    "type": "object",
+                    "required": ["public", "private"],
+                    "properties": {
+                        "public": {
+                            "type": "string"
+                        },
+                        "private": {
+                            "type": "string"
+                        },
+                    },
+                },
+            },
+        },
+        "additionalProperties": "true",
+    }
+
+    v = Draft4Validator(schema, format_checker=FormatChecker())
+    v.validate(cluster_config)
+
+
+class BenchmarkCluster:
+    """
+    Cluster for benchmark performance.
+
+    :ivar IPAddress control_node_address: IP address for control service.
+    :ivar control_service_factory: Callable taking a reactor parameter,
+        and returning a IFlockerAPIV1Client.
+    :ivar dict[IPAddress: IPAddress] public_addresses: mapping of
+        internal cluster IP addresses to public IP addresses.
+    """
+
+    def __init__(
+        self, control_node_address, control_service_factory, public_addresses
+    ):
+        self._control_node_address = control_node_address
+        self._control_service_factory = control_service_factory
+        self._public_addresses = public_addresses
+
+        self._control_service = None
+
+    @classmethod
+    def from_acceptance_test_env(cls, env):
+        """
+        Create a cluster from acceptance test environment variables.
+
+        :param dict env: Dictionary mapping acceptance test environment names
+            to values.
+        :return: A ``BenchmarkCluster`` instance.
```

Please document the exceptions this will raise given invalid data.